src/endpoints/mod_feedback.rs

```diff
@@ -0,0 +1,115 @@
+use actix_web::{get, post, web, HttpResponse, Responder};
+use serde::{Deserialize};
+use sqlx::Acquire;
+
+use crate::{
+    extractors::auth::Auth,
+    types::{
+        api::{ApiError, ApiResponse},
+        models::{
+            developer::{Developer},
+        },
+    },
+    AppData
+};
+use crate::types::models::mod_version::ModVersion;
+use crate::types::models::mod_feedback::{ModFeedback,FeedbackTypeEnum};
+
+#[derive(Deserialize)]
+pub struct GetModFeedbackPath {
+    id: String,
+    version: String
+}
+
+#[derive(Deserialize)]
+pub struct PostModFeedbackPayload {
+    feedback_type: FeedbackTypeEnum,
+    feedback: String,
+}
+
+#[get("/v1/mods/{id}/versions/{version}/feedback")]
+pub async fn get_mod_feedback(
+    data: web::Data<AppData>,
+    path: web::Path<GetModFeedbackPath>,
+    auth: Auth,
+) -> Result<impl Responder, ApiError> {
+    let dev = auth.developer()?;
+    let mut pool = data.db.acquire().await.or(Err(ApiError::DbAcquireError))?;
+
+    if !Developer::has_access_to_mod(dev.id, &path.id, &mut pool).await? && !dev.admin && !dev.verified {
+        return Err(ApiError::Forbidden);
+    }
+
+    let mut note_only = false;
+    if !Developer::has_access_to_mod(dev.id, &path.id, &mut pool).await? && !dev.admin {
+        note_only = true;
+    }
+
+    let mod_version = {
+        // latest bugs for some reason
+
+        //if path.version == "latest" {
+        //    ModVersion::get_latest_for_mod(&path.id, None, vec![], None, &mut pool).await?
+        //} else {
+        ModVersion::get_one(path.id.strip_prefix('v').unwrap_or(&path.id), &path.version, false, false, &mut pool).await?
+        //}
+    };
+
+    let feedback = ModFeedback::get_for_mod_version_id(&mod_version, note_only, &mut pool).await?;
+
+    Ok(web::Json(ApiResponse {
+        error: "".to_string(),
+        payload: feedback,
+    }))
+}
+
+#[post("/v1/mods/{id}/versions/{version}/feedback")]
+pub async fn post_mod_feedback(
+    data: web::Data<AppData>,
+    path: web::Path<GetModFeedbackPath>,
+    payload: web::Json<PostModFeedbackPayload>,
+    auth: Auth,
+) -> Result<impl Responder, ApiError> {
+    let dev = auth.developer()?;
+    let mut pool = data.db.acquire().await.or(Err(ApiError::DbAcquireError))?;
+    let mut transaction = pool.begin().await.or(Err(ApiError::TransactionError))?;
+
+    if (!dev.verified && !dev.admin) {
+        return Err(ApiError::Forbidden);
+    }
+
+    if Developer::has_access_to_mod(dev.id, &path.id, &mut transaction).await? && payload.feedback_type != FeedbackTypeEnum::Note {
```

Comment:
you should probably only call this once, and reuse the result, since every call is a db hit.

Reply:
Got it.